### ISSUE TITLE
Rule idea: color-no-non-variables

### ISSUE THREAD

**Issue by davidtheclark** [kristerkari/stylelint-scss#4]
Someone may want to enforce that any color values are named and reused as variables.

This can make the code more legible and DRY.

```scss
// bad
.foo { color: pink; background: #fff; }

// good
$some-color: pink;
$some-background: #fff;
.foo { color: $some-color; background: $some-backgorund; }
```

**Comment by kristerkari** [comment]
Should this be done in simple way so that it would only check that the values are variables, or should it also check that used variables are actually colors?

**Comment by davidtheclark** [comment]
I think this rule would actually be kind of difficult, because it would require figuring out how to check that a given value *should* be a color, then seeing if it is a variable or not. Maybe it would have to be limited in that respect.

I was just thinking that it would check that you use variables where colors are expected. One problem with trying to check the variable's value is that people often setup stylesheets in such a way that when stylelint looks at a given stylesheet it will not have access to the variable values (e.g. they're in a separate `variables.scss` file).

**Comment by kristerkari** [comment]
True. I have seen variables in different file many times, so checking for its value would not really work.

**Comment by dryoma** [comment]
I think checking variables values is a more of job of Sass compiler.

As for the rule idea: SCSS-lint (sorry to bring it up again; can't help it, since it's a functioning tool and the one seems the most popuar) hase [VariableForProperty](https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md#variableforproperty) rule. It does almost the same, except it forces the policy not on *value types* (like @davidtheclark proposed), but for *property names*.

I just thought that in the future users would want something similar to `color-no-non-variable`, but for fonts and other types of values. Wouldn't it be less redundant to make a general `value-no-non-variable` and make it accept an array of value types/property names?

**Comment by kristerkari** [comment]
@dryoma could you open a new Github issue with your idea? It's easier to track things that way :)

**Comment by dryoma** [comment]
Done.

**Comment by filip-savic** [comment]
@davidtheclark If you have variables in a `variables.scss` or similar file, you could set

```
"color-no-hex": true,
 "color-named": "never",
```

so you can't have either `color: #fff` or `color: white`, and just disable linting for those values where they are set.

**Comment by mathroc** [comment]
@filip-savic there’s also rgb/rgba ans hsl/hsla

**Comment by levacic** [comment]
I use the following plugin for that purpose: https://github.com/sh-waqar/stylelint-declaration-use-variable - though I guess it might make sense for that rule to be part of this package.

**Comment by rob-bar** [comment]
> I use the following plugin for that purpose: https://github.com/sh-waqar/stylelint-declaration-use-variable - though I guess it might make sense for that rule to be part of this package.
> 
> EDIT: I configure that rule like this:
> 
> ```
>     'sh-waqar/declaration-use-variable': [
>       [
>         'background-color',
>         'color',
>       ],
>     ],
> ```
> 
> I don't need it to be too comprehensive, as the basic sanity check done by that package is good enough for me.
> 
> Also, regarding the conversation about actually confirming that the variable _is_ a color (or something else, if a more generic rule was made) - that might even be impossible in some cases, for example a library that provides a mixin which has some kind of a `$color` parameter - you can't know at the point of the mixin definition what an end-user might pass to the mixin.

This is insufficient...
What about border-color, box-shadow, etc. etc...
The plugin allows adding more properties and even a /color/ regex sort of idea, but this will always be insufficient and a problem with future color properties. 

**There should just be a rule that checks on any color setting other than a variable. Then that should be flagged. Not caring about the property.**

**Comment by rob-bar** [comment]
I know the node-sass linter has this option. I love that. Because then I'm sure all my colors rely on the color scheme vars. And optionally I can dissable the rule where I need to per line. when I need a specific hex value.

**Comment by dartess** [comment]
> node-sass linter has this option
@rob-bar please give a link for this

**Comment by rob-bar** [comment]
These are some links:
[no-color-hex](https://github.com/sasstools/sass-lint/blob/master/docs/rules/no-color-hex.md)
[no-color-keywords](https://github.com/sasstools/sass-lint/blob/master/docs/rules/no-color-keywords.md)
[no-color-literals](https://github.com/sasstools/sass-lint/blob/master/docs/rules/no-color-literals.md)

[variable-for-property](https://github.com/sasstools/sass-lint/blob/master/docs/rules/variable-for-property.md)

**Comment by nicothin** [comment]
> I know the node-sass linter has this option. I love that. Because then I'm sure all my colors rely on the color scheme vars. And optionally I can dissable the rule where I need to per line. when I need a specific hex value.

Last stable release in October 2017 :( The project is not maintained.

**Comment by nicothin** [comment]
https://github.com/sh-waqar/stylelint-declaration-use-variable

**Comment by vmohir** [comment]
Use this plugin: https://github.com/AndyOGo/stylelint-declaration-strict-value

This is my configuration (may not be complete yet but is a starter):

```
'scale-unlimited/declaration-strict-value': [
      ['/color$/', 'z-index', 'font-size'],
      {
        expandShorthand: true,
        ignoreValues: [
          'currentColor',
          'inherit',
          'transparent',
          '/^#([0-9a-f])\\1\\1(\\1\\1\\1)?$/' /* Like #ccc */,
          '/^#([0-9a-f])([0-9a-f])\\1\\2\\1\\2$/' /* Like #c3c3c3 */,
        ],
        message:
          'Avoid using magical values like "${value}" for "${property}". Instead, use predefined variables like `var(--primary)`',
      },
    ],
```
